File: ui/desktop/src/utils/tunnel.ts
import log from './logger';

export type TunnelInfo =
  | { state: 'idle' }
  | { state: 'starting' }
  | { state: 'running'; url: string; hostname: string; secret: string }
  | { state: 'error'; error: string };

/**
 * Start the tunnel via Rust API and remember it
 */
export async function startTunnel(baseUrl: string, serverSecret: string): Promise<TunnelInfo> {
  log.info(`Starting tunnel via Rust API at ${baseUrl}`);

  const response = await fetch(`${baseUrl}/api/tunnel/start`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-Secret-Key': serverSecret,
    },
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Failed to start tunnel: ${response.statusText} - ${errorText}`);
  }

  const info: TunnelInfo = await response.json();

  log.info('Tunnel started successfully:', info);
  return info;
}

/**
 * Stop the tunnel via Rust API, and remember it
 */
export async function stopTunnel(baseUrl: string, secret: string): Promise<void> {
  log.info('Stopping tunnel via Rust API');

  try {
    const response = await fetch(`${baseUrl}/api/tunnel/stop`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Secret-Key': secret,
      },
    });

    if (!response.ok) {
      log.error(`Failed to stop tunnel: ${response.statusText}`);
    } else {
      log.info('Tunnel stopped successfully');
    }
  } catch (error) {
    log.error('Error stopping tunnel:', error);
  }
}

/**
 * Get tunnel info from Rust API
 * Note: Rust backend is the source of truth
 */
export async function syncTunnelStatus(baseUrl: string, secret: string): Promise<TunnelInfo> {
  const response = await fetch(`${baseUrl}/api/tunnel/status`, {
    headers: {
      'X-Secret-Key': secret,
    },
  });

  if (!response.ok) {
    throw new Error(`Failed to get tunnel status: ${response.statusText}`);
  }

  return await response.json();
}

File: crates/goose-server/src/routes/tunnel.rs
use crate::state::AppState;
use axum::{
    extract::State,
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{get, post},
    Json, Router,
};
use serde::Serialize;
use std::sync::Arc;
use utoipa::ToSchema;

#[derive(Debug, Serialize, ToSchema)]
pub struct ErrorResponse {
    pub error: String,
}

/// Start the tunnel
#[utoipa::path(
    post,
    path = "/api/tunnel/start",
    responses(
        (status = 200, description = "Tunnel started successfully", body = TunnelStatus),
        (status = 400, description = "Bad request", body = ErrorResponse),
        (status = 500, description = "Internal server error", body = ErrorResponse)
    )
)]
#[axum::debug_handler]
pub async fn start_tunnel(State(state): State<Arc<AppState>>) -> Response {
    match state.tunnel_manager.start().await {
        Ok(_info) => {
            let status = state.tunnel_manager.get_status().await;
            (StatusCode::OK, Json(status)).into_response()
        }
        Err(e) => {
            tracing::error!("Failed to start tunnel: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ErrorResponse {
                    error: e.to_string(),
                }),
            )
                .into_response()
        }
    }
}

/// Stop the tunnel
///
/// Stops the currently running tunnel and optionally clears the auto-start setting.
#[utoipa::path(
    post,
    path = "/api/tunnel/stop",
    responses(
        (status = 200, description = "Tunnel stopped successfully"),
        (status = 500, description = "Internal server error", body = ErrorResponse)
    )
)]
pub async fn stop_tunnel(State(state): State<Arc<AppState>>) -> Response {
    state.tunnel_manager.stop(true).await;
    StatusCode::OK.into_response()
}

/// Get tunnel status
///
/// Returns the current tunnel state and connection information if running.
#[utoipa::path(
    get,
    path = "/api/tunnel/status",
    responses(
        (status = 200, description = "Tunnel status", body = TunnelStatus)
    )
)]
pub async fn get_tunnel_status(State(state): State<Arc<AppState>>) -> Response {
    let status = state.tunnel_manager.get_status().await;
    (StatusCode::OK, Json(status)).into_response()
}

pub fn routes(state: Arc<AppState>) -> Router {
    Router::new()
        .route("/api/tunnel/start", post(start_tunnel))
        .route("/api/tunnel/stop", post(stop_tunnel))
        .route("/api/tunnel/status", get(get_tunnel_status))
        .with_state(state)
}
